### DISCUSSION TITLE
code logic for reader.go

### DISCUSSION THREAD

**Discussion by SamYuan1990** [sustainable-computing-io/kepler#273] [Q&A] [answered]
https://github.com/sustainable-computing-io/kepler/blob/b941ce513e8495efdc11e3578de0ac873e3616ea/pkg/collector/reader.go#L255-L366

1. lock.lock() not follow with defer lock.unlock()
2. this read() function seems been invoked only once? hence may I know the reason for reset() functions?
3. for loop logic as round robin, should we consider a go routine base on pod id? (each pod id as a key and a go routine for update?)
4. gpu device decouple from current logic? in case on no gpu?

**Comment by SamYuan1990** [comment]
@rootfs , what do you think?

**Reply by rootfs** [reply]
the code block is called in a go routine triggered by timer. The reset() function is to clean up the ebpf table before first entry and after each read. 

A per pod id based go routine will not help here since it is not known which pod will be collected by ebpf in the kernel. Once the table is filled by ebpf, the userspace read is quick enough (i tested it on my laptop). Having a dedicated go routine for each pod id doesn't improve the performance but it will add resources to the processing. 

gpu device is associated with each pod, how do you want to separate from the current logic?

**Reply by SamYuan1990** [reply]
so far if there no gpu device, the code still go through with gpu logic for example: `podGPUDelta`
as we have `*enableGPU`, if `GPU` is disabled, can we skip gpu related logic?

**Reply by rootfs** [reply]
sounds a good PR candidate

**Comment by marceloamaral** [comment, marked as answer]
> lock.lock() not follow with defer lock.unlock()

We unlock in the end and there is no branch in between, so we don't need the defer

> this read() function seems been invoked only once? hence may I know the reason for reset() functions?

Yes. We have an infinity loop reading the data, as expected from this function
As @rootfs said before, we must reset the counters every time that we will read the data from eBPF

> for loop logic as round robin, should we consider a go routine base on pod id? (each pod id as a key and a go routine for update?)

We don't need to create a go routine for each pod. 

> gpu device decouple from current logic? in case on no gpu?

I think it is better to keep everything together
